examples/cndpfwd/main.c

```diff
@@ -147,6 +147,56 @@ _fwd_test(jcfg_lport_t *lport, struct fwd_info *fwd)
     return n_pkts;
 }
 
+static int
+_l3fwd_test(jcfg_lport_t *lport, struct fwd_info *fwd)
+{
+    struct fwd_port *pd                          = lport->priv_;
+    struct create_txbuff_thd_priv_t *thd_private = pd->thd->priv_;
+    txbuff_t **txbuff;
+    int i, n_pkts;
+    struct ether_addr eaddr;
+    uint16_t tx_port;
+
+    if (!pd)
+        CNE_ERR_RET("fwd_port passed in lport private data is NULL\n");
+
+    txbuff = thd_private->txbuffs;
+
+    n_pkts = __rx_burst(fwd->pkt_api, pd, pd->rx_mbufs, fwd->burst);
+    if (n_pkts == PKTDEV_ADMIN_STATE_DOWN)
+        return -1;
+
+    for (i = 0; i < n_pkts; i++) {
+
+        struct cne_ipv4_hdr *rx_ip_hdr = pktmbuf_mtod_offset(pd->rx_mbufs[i], struct cne_ipv4_hdr *, ETHER_HDR_LEN);
+        rx_ip_hdr->time_to_live--;
+        rx_ip_hdr->hdr_checksum = htons(ntohs(rx_ip_hdr->hdr_checksum) + 1);
+        uint32_t ip_addr = ntohl(rx_ip_hdr->dst_addr);
+
+        l3fwd_fib_lookup(&ip_addr, &eaddr, &tx_port);
+
+        jcfg_lport_t *dst = jcfg_lport_by_index(fwd->jinfo, tx_port);
+
+        MAC_REWRITE(pktmbuf_mtod(pd->rx_mbufs[i], void *), &eaddr);
+
+        (void)txbuff_add(txbuff[dst->lpid], pd->rx_mbufs[i]);
```

Comment:
Just noticed that we lost our logic here that should check whether the `jcfg_lport_by_index` call succeeded, which means this line could try to dereference NULL if the user specifies a nonexistent TX port index in their FIB definition.

The regular `fwd` test echoes packets back out of the RX port if there's no TX port matching the low byte of the MAC, but in our case, since the TX port indices are defined ahead of time, I'm inclined to drop them...

Reply:
If we drop the packets with invalid TX port then we can not tell where the packet is dropped. Sending the packet back out the RX port seemed like a reasonable solution as we can see the packets getting looped back to the RX port. I can see both ways could work, let's drop them for now. Maybe some type of indicator to tell the user what is happening as it is an error.

When we add the jsonc **fib-rules**, we need to check the TX port exists and if not error out the rule.

The check for **dst** being null should be done, with a message could work. It could a be a huge number of messages, but I think it is reasonable.